### ISSUE TITLE
AGREE thinks a data definitions file is a subcomponent in need of an AGREE contract

### ISSUE THREAD

**Issue by jendavis** [loonwerks/AGREE#31]
See the example below (two files). When running analysis on Top.impl, I get the following error message:
"Variable 'Types_Constants.Response_Agree' appears in an assertion, lemma or equation statement in component 'Top_impl_Instance' but subcomponent 'Types_Constants' does not contain an AGREE annex"

However, Types_Constants is not a subcomponent. It is an AADL data definitions file and should not be required to have an AGREE annex. Note that an enumeration type is defined in the Types_Constants package.

~~~
package Types_Constants
public
	with Base_Types;
	with Data_Model;

	-------------------------------------
	-- Data Types
	-------------------------------------

	data Response
		properties
			Data_Model::Data_Representation => Enum;
			Data_Model::Enumerators => ("Agree", "Disagree", "Neutral");
	end Response;
	
	data H2I
	end H2I;
	
	data implementation H2I.impl
		subcomponents
			Sensor1_Unreliable_Response: data Response;
			Sensor2_Unreliable_Response: data Response;
			Sensor3_Unreliable_Response: data Response;			
	end H2I.impl;
	
	data I2H
	end I2H;
	
	data implementation I2H.impl
		subcomponents
			Sensor1_Reliable: 	data Base_Types::Boolean;
			Sensor2_Reliable: 	data Base_Types::Boolean;
			Sensor3_Reliable: 	data Base_Types::Boolean;
			Position: 			data Position.impl;
	end I2H.impl;
	
	data Position
	end Position;
	
	data implementation Position.impl
		subcomponents
			X: data Base_Types::Float;
			Y: data Base_Types::Float;
			Z: data Base_Types::Float;							
	end Position.impl;

	data Sensor_Data
	end Sensor_Data;
	
	data implementation Sensor_Data.impl
		subcomponents
			Position: data Position.impl;
	end Sensor_Data.impl;
	
end Types_Constants;
~~~
~~~
package Human_IAS_Team
public
	with Types_Constants;
	renames Types_Constants::all;

	-------------------------------------
	-- Human
	-------------------------------------
	
	system Human
		features
			InputFromIAS : in data port I2H.impl;
			InputFromSensor1 : in data port Sensor_Data.impl;
			InputFromSensor2 : in data port Sensor_Data.impl;
			InputFromSensor3 : in data port Sensor_Data.impl;
			Output : out data port H2I.impl;
		annex agree {**
				
			guarantee "Respond to message that Sensor1 is unreliable":
				prev(not InputFromIAS.Sensor1_Reliable, false) <=> ((Output.Sensor1_Unreliable_Response = enum(Response, Agree)) or (Output.Sensor1_Unreliable_Response = enum(Response, Disagree))); 
			
			guarantee "Respond to message that Sensor2 is unreliable":
				prev(not InputFromIAS.Sensor2_Reliable, false) <=> ((Output.Sensor2_Unreliable_Response = enum(Response, Agree)) or (Output.Sensor2_Unreliable_Response = enum(Response, Disagree)));
			
			guarantee "Respond to message that Sensor3 is unreliable":
				prev(not InputFromIAS.Sensor2_Reliable, false) <=> ((Output.Sensor3_Unreliable_Response = enum(Response, Agree)) or (Output.Sensor3_Unreliable_Response = enum(Response, Disagree)));		
				
			 	
		**};
	end Human;
	
	-------------------------------------
	-- IAS
	-------------------------------------
	
	system IAS
		features
			InputFromHuman : in data port H2I.impl;
			InputFromSensor1 : in data port Sensor_Data.impl;
			InputFromSensor2 : in data port Sensor_Data.impl;
			InputFromSensor3 : in data port Sensor_Data.impl;
			Output : out data port I2H.impl;

		annex agree {**
							
			eq Sensor1_Reliable: bool;
			eq Sensor2_Reliable: bool;
			eq Sensor3_Reliable: bool;
				
			guarantee "Sensor sensor status to pilot":
					(Output.Sensor1_Reliable = Sensor1_Reliable)
				and (Output.Sensor2_Reliable = Sensor2_Reliable)
				and (Output.Sensor3_Reliable = Sensor3_Reliable);			
									 								
		**};

	end IAS;
	
	-------------------------------------
	-- Sensors
	-------------------------------------
		
	system Sensor1
		features
			Output : out data port Position.impl;
	end Sensor1;
	
	system Sensor2
		features
			Output : out data port Position.impl;
	end Sensor2;
	
	system Sensor3
		features
			Output : out data port Position.impl;
	end Sensor3;
	
	-------------------------------------
	-- Human-Machine Team
	-------------------------------------
	
	system Top
		annex agree {**
			guarantee "Placeholder gaurantee to get AGREE to run": true;
		**};
	end Top;

	system implementation Top.impl
		subcomponents
			Human: system Human;
			IAS: system IAS;
			Sensor1: system Sensor1;
			Sensor2: system Sensor2;
			Sensor3: system Sensor3;
		connections
			Top_impl_new_connection: feature IAS.Output -> Human.InputFromIAS;
			Top_impl_new_connection2: feature Sensor1.Output -> IAS.InputFromSensor1;
			Top_impl_new_connection3: feature Sensor2.Output -> IAS.InputFromSensor2;
			Top_impl_new_connection4: feature Sensor3.Output -> IAS.InputFromSensor3;
			Top_impl_new_connection5: feature Human.Output -> IAS.InputFromHuman;
			Top_impl_new_connection6: feature Sensor1.Output -> Human.InputFromSensor1;
			Top_impl_new_connection7: feature Sensor2.Output -> Human.InputFromSensor2;
			Top_impl_new_connection8: feature Sensor3.Output -> Human.InputFromSensor3;
		annex agree{**
		
			------------------------------------------------------------------------------
			-- LEMMAS (These should be true based on the guarantees of the subcomponents.)
			------------------------------------------------------------------------------
				
			lemma "Pilot acknowledges unreliable sensor alerts": 
					(prev(not IAS.Output.Sensor1_Reliable, false) <=> ((Human.Output.Sensor1_Unreliable_Response = enum(Response, Agree)) or (Human.Output.Sensor1_Unreliable_Response = enum(Response, Disagree))))
				and (prev(not IAS.Output.Sensor2_Reliable, false) <=> ((Human.Output.Sensor2_Unreliable_Response = enum(Response, Agree)) or (Human.Output.Sensor2_Unreliable_Response = enum(Response, Disagree))))
				and (prev(not IAS.Output.Sensor3_Reliable, false) <=> ((Human.Output.Sensor3_Unreliable_Response = enum(Response, Agree)) or (Human.Output.Sensor3_Unreliable_Response = enum(Response, Disagree))));
							
		**};
	end Top.impl;
	
	
end Human_IAS_Team;

~~~

**Comment by kfhoech** [comment]
Confirmed.

AGREE translation rejects IdExpr where the id references an entity not containing an annex.  The purpose of this check is to avoid a less user-friendly message resulting from an undefined reference during the JKind analysis.  However, as implemented, the check does not correctly implement the intent.

The relevant code is [here](https://github.com/loonwerks/AGREE/blob/7ea5d27a8be32d980e13d88ec39529ee9d6b408c/com.rockwellcollins.atc.agree.analysis/src/com/rockwellcollins/atc/agree/analysis/ast/AgreeASTBuilder.java#L555-L597).

This check does nothing to implement a portion of the translation; it merely throws an AgreeException if it's not satisfied.  The check, however, is poorly conceived in that it checks too late in the process to determine whether the IdExpr refers to something that does not need an annex (e.g. AADL property or enumeration data type). Further it checks only assertions and lemmas and misses occurrences in guarantees, assumptions, equations, and other AGREE expressions.

This check should be removed entirely in favor of validator checks.

**Comment by kfhoech** [comment]
Fixed by #34.
